**This notebook is an exercise in the [Intermediate Machine Learning](https://www.kaggle.com/learn/intermediate-machine-learning) course.  You can reference the tutorial at [this link](https://www.kaggle.com/alexisbcook/missing-values).**

---


Now it's your turn to test your new knowledge of **missing values** handling. You'll probably find it makes a big difference.

# Setup

The questions will give you feedback on your work. Run the following cell to set up the feedback system.

In [1]:
# Set up code checking
import os
if not os.path.exists("../input/train.csv"):
    os.symlink("../input/home-data-for-ml-course/train.csv", "../input/train.csv")  
    os.symlink("../input/home-data-for-ml-course/test.csv", "../input/test.csv") 
from learntools.core import binder
binder.bind(globals())
from learntools.ml_intermediate.ex2 import *
print("Setup Complete")

Setup Complete


In this exercise, you will work with data from the [Housing Prices Competition for Kaggle Learn Users](https://www.kaggle.com/c/home-data-for-ml-course). 

![Ames Housing dataset image](https://i.imgur.com/lTJVG4e.png)

Run the next code cell without changes to load the training and validation sets in `X_train`, `X_valid`, `y_train`, and `y_valid`.  The test set is loaded in `X_test`.

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Read the data
X_full = pd.read_csv('../input/train.csv', index_col='Id')
X_test_full = pd.read_csv('../input/test.csv', index_col='Id')

# Remove rows with missing target, separate target from predictors
X_full.dropna(axis=0, subset=['SalePrice'], inplace=True) # axis=0 means row and axis=1 means column, inplace=True means it will change the original data. Subset means it will only drop the rows that have missing values in the SalePrice column.
y = X_full.SalePrice # y is the target, SalePrice is the column name
X_full.drop(['SalePrice'], axis=1, inplace=True) # drop the SalePrice column from the original data

# To keep things simple, we'll use only numerical predictors
X = X_full.select_dtypes(exclude=['object']) # makes a copy of the data and selects only the columns with numerical data types. Here we are excluding the object data types (which are strings).

X_test = X_test_full.select_dtypes(exclude=['object']) # makes a copy of the data and selects only the columns with numerical data types. Here we are excluding the object data types (which are strings).

# Break off validation set from training data
X_train, X_valid, y_train, y_valid = train_test_split(X, y, train_size=0.8, test_size=0.2, random_state=0)

# > My Notes
> how to possibly improv the above code improvement 
>
> proposed datapipeline:
>
> - pass y with `inplace = true` to drop non-numrical col. 
>
> in previous step `X_full.dropna(axis=0, subset=['SalePrice'], inplace=True)`, we drop rows with missing target feature however, does it drop non-numerical dat also? No. i.e.`2 billion usd` is non-numrecal

In [3]:
X_train.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
Id                                                                           
619          20         90.0    11694            9            5       2007   
871          20         60.0     6600            5            5       1962   
93           30         80.0    13360            5            7       1921   
818          20          NaN    13265            8            5       2002   
303          20        118.0    13704            7            5       2001   

     YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  ...  GarageArea  \
Id                                                     ...               
619          2007       452.0          48           0  ...         774   
871          1962         0.0           0           0  ...         308   
93           2006         0.0         713           0  ...         432   
818          2002       148.0        1218           0  ...         857   
303          2002       150.0           0           0  ...         843   

     WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  \
Id                                                                              
619           0          108              0          0          260         0   
871           0            0              0          0            0         0   
93            0            0             44          0            0         0   
818         150           59              0          0            0         0   
303         468           81              0          0            0         0   

     MiscVal  MoSold  YrSold  
Id                            
619        0       7    2007  
871        0       8    2009  
93         0       8    2009  
818        0       7    2008  
303        0       1    2006  

[5 rows x 36 columns]

You can already see a few missing values in the first several rows.  In the next step, you'll obtain a more comprehensive understanding of the missing values in the dataset.

# Step 1: Preliminary investigation

Run the code cell below without changes.

In [4]:
# Shape of training data (num_rows, num_columns)
print(X_train.shape)

# Number of missing values in each column of training data
missing_val_count_by_column = (X_train.isnull().sum()) ## isnull() returns a boolean value for each cell in the dataframe. sum() returns the number of True values in each column.
print(missing_val_count_by_column[missing_val_count_by_column > 0])

(1168, 36)
LotFrontage    212
MasVnrArea       6
GarageYrBlt     58
dtype: int64


### Part A

Use the above output to answer the questions below.

In [5]:
# Fill in the line below: How many rows are in the training data?
num_rows = X_train.shape[0] # it is 1168 because the shape is (1168, 37)
print("num_rows:", num_rows)

# Fill in the line below: How many columns in the training data
# have missing values?
num_cols_with_missing = (missing_val_count_by_column[missing_val_count_by_column > 0]).shape[0] # it is 3 because the shape is (3,). see the output of the previous cell.
print("num_cols_with_missing:", num_cols_with_missing)

# Fill in the line below: How many missing entries are contained in 
# all of the training data?
tot_missing = missing_val_count_by_column.sum() # it is 6965 because the sum of the missing values is 6965
print("tot_missing:", tot_missing)

# Check your answers
step_1.a.check()

num_rows: 1168
num_cols_with_missing: 3
tot_missing: 276


<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

In [6]:
# Lines below will give you a hint or solution code
# step_1.a.hint()
# step_1.a.solution()

### Part B
Considering your answers above, what do you think is likely the best approach to dealing with the missing values?

# > my code: 

First find, the percentage of missing values per each column of the dataset.

In [7]:
# percentage of missing values in each column of training data 
missing_val_percent_by_column = (X_train.isnull().sum()) / num_rows * 100

# print the above results using the round function and concatenate  "%" to the end of the string

print(round(missing_val_percent_by_column[missing_val_percent_by_column > 0], 1).astype(str) + "%")

LotFrontage    18.2%
MasVnrArea      0.5%
GarageYrBlt     5.0%
dtype: object


# > My notes

>The percentage of missing values in each column is shown above.  The columns are sorted in descending order by the percentage of missing values. Note that the `SalePrice` column is the target, which is not missing any values.

>The percentage of missing values in each column are low enough that you can drop the rows with missing values.  You can do this by setting the `dropna` argument to `True` in the `get_numeric_data()` method.  This will drop all the columns with missing values.  You can also drop the columns with missing values by setting the `drop` argument to `True` in the `dropna()` method.  This will drop all the rows with missing values.

>Here we are going to use imputation.  Imputation fills in the missing values with some number.  The imputed value won't be exactly right in most cases, but it usually leads to more accurate models than you would get from dropping the column entirely. We will use the `SimpleImputer` class from scikit-learn to impute the missing values.  The code below will impute the missing values with the mean value along each column.  You can read more about the `SimpleImputer` class [here](https://scikit-learn.org/stable/modules/generated/sklearn.impute.SimpleImputer.html?highlight=simpleimputer#sklearn.impute.SimpleImputer).

>We should keep in mind that the test set will have missing values in the same columns as the training set. So, we'll need to handle the missing values in the test set in the same way you handle the missing values in the training set.



In [8]:
# Check your answer (Run this code cell to receive credit!)
step_1.b.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct:</span> 

Since there are relatively few missing entries in the data (the column with the greatest percentage of missing values is missing less than 20% of its entries), we can expect that dropping columns is unlikely to yield good results.  This is because we'd be throwing away a lot of valuable data, and so imputation will likely perform better.

In [9]:
step_1.b.hint()

<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> Does the dataset have a lot of missing values, or just a few?  Would we lose much information if we completely ignored the columns with missing entries?

To compare different approaches to dealing with missing values, you'll use the same `score_dataset()` function from the tutorial.  This function reports the [mean absolute error](https://en.wikipedia.org/wiki/Mean_absolute_error) (MAE) from a random forest model.

In [10]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

# Function for comparing different approaches
def score_dataset(X_train, X_valid, y_train, y_valid):
    model = RandomForestRegressor(n_estimators=100, random_state=0)
    model.fit(X_train, y_train)
    preds = model.predict(X_valid)
    return mean_absolute_error(y_valid, preds)

# Step 2: Drop columns with missing values

In this step, you'll preprocess the data in `X_train` and `X_valid` to remove columns with missing values.  Set the preprocessed DataFrames to `reduced_X_train` and `reduced_X_valid`, respectively.  

In [11]:
# Fill in the line below: get names of columns with missing values
cols_with_missing = [col for col in X_train.columns
                        if X_train[col].isnull().any()]

print('cols_with_missing:', cols_with_missing)

# Fill in the lines below: drop columns in training and validation data
reduced_X_train = X_train.drop(cols_with_missing, axis=1)
reduced_X_valid = X_valid.drop(cols_with_missing, axis=1)

# Check your answers
step_2.check()

cols_with_missing: ['LotFrontage', 'MasVnrArea', 'GarageYrBlt']


<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

In [12]:
# Lines below will give you a hint or solution code
step_2.hint()
step_2.solution()

<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> Begin by finding the list of columns in the data with missing values.  Then, drop these columns in both the training and validation data with the `drop()` method.

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> 
```python
# Get names of columns with missing values
cols_with_missing = [col for col in X_train.columns
                     if X_train[col].isnull().any()]

# Drop columns in training and validation data
reduced_X_train = X_train.drop(cols_with_missing, axis=1)
reduced_X_valid = X_valid.drop(cols_with_missing, axis=1)

```

Run the next code cell without changes to obtain the MAE for this approach.

In [13]:
print("MAE (Drop columns with missing values):")
print(score_dataset(reduced_X_train, reduced_X_valid, y_train, y_valid))

MAE (Drop columns with missing values):
17837.82570776256


# Step 3: Imputation

### Part A

Use the next code cell to impute missing values with the mean value along each column.  Set the preprocessed DataFrames to `imputed_X_train` and `imputed_X_valid`.  Make sure that the column names match those in `X_train` and `X_valid`.

In [14]:
from sklearn.impute import SimpleImputer

# Fill in the lines below: imputation
my_imputer = SimpleImputer()
imputed_X_train = pd.DataFrame(my_imputer.fit_transform(X_train)) # impute the training data
imputed_X_valid = pd.DataFrame(my_imputer.transform(X_valid)) # impute the validation data


# Fill in the lines below: imputation removed column names; put them back
imputed_X_train.columns = X_train.columns  # add the column names back to the imputed training data
imputed_X_valid.columns = X_valid.columns # add the column names back to the imputed validation data


# Check your answers
step_3.a.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

In [15]:
# Lines below will give you a hint or solution code
#step_3.a.hint()
#step_3.a.solution()

# > my code
Displaying mean values for each column where there are missing values:

In [16]:
print("LotFrontage mean:", X_train['LotFrontage'].mean())
print("MasVnrArea mean:", X_train['MasVnrArea'].mean())
print("GarageYrBlt mean:", X_train['GarageYrBlt'].mean())

LotFrontage mean: 69.61401673640168
MasVnrArea mean: 103.48106712564544
GarageYrBlt mean: 1978.1405405405405


Run the next code cell without changes to obtain the MAE for this approach.

In [17]:
print("MAE (Imputation):")
print(score_dataset(imputed_X_train, imputed_X_valid, y_train, y_valid))

MAE (Imputation):
18062.894611872147


### Part B

Compare the MAE from each approach.  Does anything surprise you about the results?  Why do you think one approach performed better than the other?

# > my code
Displaying mean values for each column where there are missing values:

In [18]:
MAE_imputed = score_dataset(imputed_X_train, imputed_X_valid, y_train, y_valid)
print("MAE (Imputation):")
print(MAE_imputed)
      
MAE_reduced = score_dataset(reduced_X_train, reduced_X_valid, y_train, y_valid)
print("MAE (Reduced):")
print(score_dataset(reduced_X_train, reduced_X_valid, y_train, y_valid))


MAE (Imputation):
18062.894611872147
MAE (Reduced):
17837.82570776256


In [19]:
# MAE (Imputation) - MAE (Drop columns with missing values)
print("MAE (Imputation) - MAE (Drop columns with missing values):")
print(score_dataset(imputed_X_train, imputed_X_valid, y_train, y_valid) - score_dataset(reduced_X_train, reduced_X_valid, y_train, y_valid))

MAE (Imputation) - MAE (Drop columns with missing values):
225.06890410958658


# > my notes

> The MAE (Dropping columns with missing values) is 17837.83, while the MAE (Imputation) is 18062.89.  The MAE from dropping columns is lower, so this approach appears to give more accurate models.  However, the difference in MAE isn't large, so your choice of approach will probably depend on what you're trying to do. If speed is important, then dropping columns will be faster.  If missing values are rare, then dropping rows might be reasonable.  If most values are missing, then imputation will probably give more accurate models. And if some values are missing at random, then imputation will probably give more accurate models.

In [20]:
# Check your answer (Run this code cell to receive credit!)
step_3.b.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct:</span> 

Given that thre are so few missing values in the dataset, we'd expect imputation to perform better than dropping columns entirely.  However, we see that dropping columns performs slightly better!  While this can probably partially be attributed to noise in the dataset, another potential explanation is that the imputation method is not a great match to this dataset.  That is, maybe instead of filling in the mean value, it makes more sense to set every missing value to a value of 0, to fill in the most frequently encountered value, or to use some other method.  For instance, consider the `GarageYrBlt` column (which indicates the year that the garage was built).  It's likely that in some cases, a missing value could indicate a house that does not have a garage.  Does it make more sense to fill in the median value along each column in this case?  Or could we get better results by filling in the minimum value along each column?  It's not quite clear what's best in this case, but perhaps we can rule out some options immediately - for instance, setting missing values in this column to 0 is likely to yield horrible results!

In [21]:
step_3.b.hint()

<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> Did removing missing values yield a larger or smaller MAE than imputation? Does this agree with the coding example from the tutorial?

# Step 4: Generate test predictions

In this final step, you'll use any approach of your choosing to deal with missing values.  Once you've preprocessed the training and validation features, you'll train and evaluate a random forest model.  Then, you'll preprocess the test data before generating predictions that can be submitted to the competition!

### Part A

Use the next code cell to preprocess the training and validation data.  Set the preprocessed DataFrames to `final_X_train` and `final_X_valid`.  **You can use any approach of your choosing here!**  in order for this step to be marked as correct, you need only ensure:
- the preprocessed DataFrames have the same number of columns,
- the preprocessed DataFrames have no missing values, 
- `final_X_train` and `y_train` have the same number of rows, and
- `final_X_valid` and `y_valid` have the same number of rows.

In [22]:
# Preprocessed training and validation features
final_X_train = imputed_X_train # imputed training data
final_X_valid = imputed_X_valid # imputed validation data

# Check your answers
step_4.a.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

In [23]:
# Lines below will give you a hint or solution code
#step_4.a.hint()
#step_4.a.solution()

Run the next code cell to train and evaluate a random forest model.  (*Note that we don't use the `score_dataset()` function above, because we will soon use the trained model to generate test predictions!*)

In [24]:
# Define and fit model
model = RandomForestRegressor(n_estimators=100, random_state=0)
model.fit(final_X_train, y_train)

# Get validation predictions and MAE
preds_valid = model.predict(final_X_valid)
print("MAE (Your approach):")
print(mean_absolute_error(y_valid, preds_valid))

MAE (Your approach):
18062.894611872147


### Part B

Use the next code cell to preprocess your test data.  Make sure that you use a method that agrees with how you preprocessed the training and validation data, and set the preprocessed test features to `final_X_test`.

Then, use the preprocessed test features and the trained model to generate test predictions in `preds_test`.

In order for this step to be marked correct, you need only ensure:
- the preprocessed test DataFrame has no missing values, and
- `final_X_test` has the same number of rows as `X_test`.

In [25]:
# Fill in the line below: preprocess test data
final_X_test = pd.DataFrame(my_imputer.transform(X_test)) # impute the test data

# Fill in the line below: get test predictions
preds_test = model.predict(final_X_test) # get test predictions

# Check your answers
step_4.b.check()

/opt/conda/lib/python3.7/site-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  "X does not have valid feature names, but"


<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

# > My notes

```
X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(
```

>The above warning is expected.  The test set does not have feature names, so the model can't check that the test set has the same features as the training set.  However, the model will still generate predictions. You can read more about this warning [here](https://github.com/tylerjrichards/Getting-Started-with-Streamlit-for-Data-Science/issues/5). Read more about feature names [here](https://www.kaggle.com/dansbecker/using-categorical-data-with-one-hot-encoding). 
>
>We can ignore this warning for now.


    

In [26]:
# Lines below will give you a hint or solution code
step_4.b.hint()
step_4.b.solution()


<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> After preprocessing the test data, you can get the model's predictions by using `model.predict()`.

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> 
```python
# Preprocess test data
final_X_test = pd.DataFrame(final_imputer.transform(X_test))

# Get test predictions
preds_test = model.predict(final_X_test)

```

Run the next code cell without changes to save your results to a CSV file that can be submitted directly to the competition.

In [27]:
# Save test predictions to file
output = pd.DataFrame({'Id': X_test.index,
                       'SalePrice': preds_test})
output.to_csv('submission.csv', index=False)

# Submit your results

Once you have successfully completed Step 4, you're ready to submit your results to the leaderboard!  (_You also learned how to do this in the previous exercise.  If you need a reminder of how to do this, please use the instructions below._)  

First, you'll need to join the competition if you haven't already.  So open a new window by clicking on [this link](https://www.kaggle.com/c/home-data-for-ml-course).  Then click on the **Join Competition** button.

![join competition image](https://i.imgur.com/wLmFtH3.png)

Next, follow the instructions below:
1. Begin by clicking on the **Save Version** button in the top right corner of the window.  This will generate a pop-up window.  
2. Ensure that the **Save and Run All** option is selected, and then click on the **Save** button.
3. This generates a window in the bottom left corner of the notebook.  After it has finished running, click on the number to the right of the **Save Version** button.  This pulls up a list of versions on the right of the screen.  Click on the ellipsis **(...)** to the right of the most recent version, and select **Open in Viewer**.  This brings you into view mode of the same page. You will need to scroll down to get back to these instructions.
4. Click on the **Output** tab on the right of the screen.  Then, click on the file you would like to submit, and click on the **Submit** button to submit your results to the leaderboard.

You have now successfully submitted to the competition!

If you want to keep working to improve your performance, select the **Edit** button in the top right of the screen. Then you can change your code and repeat the process. There's a lot of room to improve, and you will climb up the leaderboard as you work.


# Keep going

Move on to learn what **[categorical variables](https://www.kaggle.com/alexisbcook/categorical-variables)** are, along with how to incorporate them into your machine learning models.  Categorical variables are very common in real-world data, but you'll get an error if you try to plug them into your models without processing them first!

---




*Have questions or comments? Visit the [course discussion forum](https://www.kaggle.com/learn/intermediate-machine-learning/discussion) to chat with other learners.*